registry/warming.go

```diff
@@ -0,0 +1,234 @@
+			if !withinExpiryBuffer(expiry, refreshWhenExpiryWithin) {
+				continue
+			}
+			// If we're within the expiry buffer, we need to update quick!
+			expired = true
+		}
+		toUpdate = append(toUpdate, i)
+	}
+
+	if len(toUpdate) == 0 {
+		return
+	}
+
+	if expired {
+		w.Logger.Log("expiring", id.HostNamespaceImage())
+	}
+
+	// Now refresh the manifests for each tag (in lots of goroutines for improved performance)
+	toFetch := make(chan flux.ImageID, len(toUpdate))
+	fetched := make(chan flux.Image, len(toUpdate))
+	for i := 0; i < w.Burst; i++ {
+		go func() {
+			for i := range toFetch {
+				// Get the image from the remote
+				img, err := client.Manifest(i)
+				if err != nil {
+					if !strings.Contains(err.Error(), context.DeadlineExceeded.Error()) && !strings.Contains(err.Error(), "net/http: request canceled") {
+						w.Logger.Log("err", errors.Wrap(err, "requesting manifests"))
+					}
+				}
+				fetched <- img // Always return an image, otherwise the for loop below will never finish.
+			}
+		}()
+	}
+	for _, img := range toUpdate {
+		toFetch <- img
+	}
+	close(toFetch)
+
+	// Write received manifests back to memcache
+	for i := 0; i < cap(fetched); i++ {
+		img := <-fetched
+		if img.ID.String() == "" {
+			continue
+		}
+		key, err := cache.NewManifestKey(username, img.ID)
+		if err != nil {
+			w.Logger.Log("err", errors.Wrap(err, "creating key for memcache"))
+			continue
+		}
+		// Write back to memcache
+		val, err := json.Marshal(img)
+		if err != nil {
+			w.Logger.Log("err", errors.Wrap(err, "serializing tag to store in cache"))
+			return
+		}
+		err = w.Writer.SetKey(key, val)
+		if err != nil {
+			w.Logger.Log("err", errors.Wrap(err, "storing manifests in cache"))
+			return
+		}
+	}
+	close(fetched)
+	w.Logger.Log("updated", id.HostNamespaceImage())
+}
+
+func withinExpiryBuffer(expiry time.Time, buffer time.Duration) bool {
+	// if the `time.Now() + buffer  > expiry`,
+	// then we're within the expiry buffer
+	if time.Now().Add(buffer).After(expiry) {
+		return true
+	}
+	return false
+}
+
+// Queue provides an updating repository queue for the warmer.
+// If no items are added to the queue this will randomly add a new
+// registry to warm
+type Queue struct {
+	RunningContainers    func() []flux.ImageID
+	Logger               log.Logger
+	RegistryPollInterval time.Duration
+	warmQueue            chan flux.ImageID
+	queueLock            sync.Mutex
+}
+
+func NewQueue(runningContainersFunc func() []flux.ImageID, l log.Logger, emptyQueueTick time.Duration) Queue {
+	return Queue{
+		RunningContainers:    runningContainersFunc,
+		Logger:               l,
+		RegistryPollInterval: emptyQueueTick,
+		warmQueue:            make(chan flux.ImageID, 100), // Don't close this. It will be GC'ed when this instance is destroyed.
+	}
+}
+
+// Queue loop to maintain the queue and periodically add a random
+// repository that is running in the cluster.
+func (w *Queue) Loop(stop chan struct{}, wg *sync.WaitGroup) {
```

Comment:
IMO the way this should work is that periodically (say once a minute) we enumerate all the image IDs and feed them into the warm queue. If the queue becomes full, we should block, rather than hammering k8s asking it to enumerate things we can't use anyway.